ignite/pkg/cosmosgen/generate.go

```diff
@@ -124,118 +153,189 @@ func (g *generator) setup(ctx context.Context) (err error) {
-	if err != nil {
-		return err
-	}
-
-	dep, found := filterCosmosSDKModule(g.deps)
-	if !found {
-		return ErrMissingSDKDep
-	}
-
-	// Find the full path to the Cosmos SDK Go package.
-	// The path is required to be able to discover proto packages for the
-	// set of "cosmossdk.io" packages that doesn't contain the proto files.
-	g.sdkDir, err = gomodule.LocatePath(ctx, g.cacheStorage, g.appPath, dep)
-	if err != nil {
-		return err
-	}
+	return err
+}
 
-	// Go through the Go dependencies of the user's app within go.mod, some of them
-	// might be hosting Cosmos SDK modules that could be in use by user's blockchain.
-	//
-	// Cosmos SDK is a dependency of all blockchains, so it's absolute that we'll be
-	// discovering all modules of the SDK as well during this process.
-	//
-	// Even if a dependency contains some SDK modules, not all of these modules could
-	// be used by user's blockchain. This is fine, we can still generate TS clients
-	// for those non modules, it is up to user to use (import in typescript) not use
-	// generated modules.
-	//
-	// TODO: we can still implement some sort of smart filtering to detect non used
-	// modules by the user's blockchain at some point, it is a nice to have.
+func (g *generator) processThirdPartyModules(ctx context.Context) error {
 	moduleCache := cache.New[protoAnalysis](g.cacheStorage, moduleCacheNamespace)
-	for _, dep := range g.deps {
-		// Try to get the cached list of modules for the current dependency package
-		cacheKey := cache.Key(dep.Path, dep.Version)
-		depInfo, err := moduleCache.Get(cacheKey)
-		if err != nil && !errors.Is(err, cache.ErrorNotFound) {
-			return err
-		}
+	sdkModuleCache := cache.New[protoAnalysis](g.cacheStorage, sdkModuleCacheNamespace)
 
-		// Discover the modules of the dependency package when they are not cached
-		if errors.Is(err, cache.ErrorNotFound) {
-			// Get the absolute path to the package's directory
-			path, err := gomodule.LocatePath(ctx, g.cacheStorage, g.appPath, dep)
-			if err != nil {
-				return err
-			}
+	// Process dependencies in parallel for better performance
+	type depResult struct {
+		path     string
+		analysis protoAnalysis
+		err      error
+	}
 
-			// Discover any modules defined by the package.
-			// Use an empty string for proto directory because it will be
-			// discovered automatically within the dependency package path.
-			modules, err := module.Discover(ctx, g.appPath, path, module.WithSDKDir(g.sdkDir))
-			if err != nil {
-				return err
-			}
+	results := make(chan depResult, len(g.deps))
+	semaphore := make(chan struct{}, 5) // Limit concurrent operations
 
-			// Dependency/includes resolution per module is done to solve versioning issues
-			var (
-				includes  protoIncludes
-				cacheable = true
-			)
-			if len(modules) > 0 {
-				includes, cacheable, err = g.resolveIncludes(ctx, path, defaults.ProtoDir)
-				if err != nil {
-					return err
+	for _, dep := range g.deps {
+		go func(dep gomodule.Version) {
```

Comment:
IMHO, we should use a controlled go routine using the context so that the routine can die with the context, like the errgroup

Reply:
good one, an early exist is most welcome!